rclcpp/include/rclcpp/executors/executor_entities_collection.hpp

```diff
@@ -72,34 +73,47 @@ void update_entities(
   std::function<void(const typename CollectionType::EntitySharedPtr &)> on_removed
 )
 {
-  for (auto it = update_to.begin(); it != update_to.end(); ) {
+  for (auto it = update_to.begin_ordered(); it != update_to.end_ordered(); ) {
     if (update_from.count(it->first) == 0) {
       auto entity = it->second.entity.lock();
       if (entity) {
         on_removed(entity);
       }
-      it = update_to.erase(it);
+      it = update_to.erase_ordered(it);
     } else {
       ++it;
     }
   }
-  for (auto it = update_from.begin(); it != update_from.end(); ++it) {
+  for (auto it = update_from.begin_ordered(); it != update_from.end_ordered(); ++it) {
     if (update_to.count(it->first) == 0) {
       auto entity = it->second.entity.lock();
       if (entity) {
         on_added(entity);
       }
-      update_to.insert(*it);
+      bool inserted = update_to.insert(*it);
+      // Should never be false, so this is a defensive check, mark unused too
+      // in order to avoid a warning in release builds.
+      assert(inserted);
+      RCUTILS_UNUSED(inserted);
     }
   }
 }
 
 /// A collection of entities, indexed by their corresponding handles
 template<typename EntityKeyType, typename EntityValueType>
 class EntityCollection
-  : public std::unordered_map<const EntityKeyType *, CollectionEntry<EntityValueType>>
 {
 public:
+  /// Type of the map used for random access
+  using MapType = std::unordered_map<const EntityKeyType *, CollectionEntry<EntityValueType>>;
+
+  /// Type of the vector for insertion order access
+  // Note, we cannot use typename MapType::value_type because it makes the first
+  // item in the pair const, which prevents copy assignment of the pair, which
+  // prevents std::vector::erase from working later...
+  using VectorType = std::vector<std::pair<const EntityKeyType *,
+      CollectionEntry<EntityValueType>>>;
```

Comment:
This detail was annoying, the MapType::value_type makes the first argument of the pair const (would be `const EntityKeyType * const` in this case), which prevents the use of the `operator=` on the pair, which prevents `std::vector::erase` from working...

Anyway, this is essentially a `std::vector<typename MapType::value_type>` except for that detail...

Reply:
Why not just store the iterator returned by std::map::insert ?

Reply:
Because they can be invalidated by future insertions.

Reply:
I also considered storing the pointer (which is the key for the map) only, but then when iterating in order you need to do a lookup in the map for each one you want to use, which is made redundant if you just store the information twice. In fact this conclusion is what several of the "ordered dict in c++" projects I looked at online ended up using.

Reply:
> Because they can be invalidated by future insertions.

Nope, not for std::map : 
https://en.cppreference.com/w/cpp/container/map/insert

No iterators or references are invalidated. If the insertion is successful, pointers and references to the element obtained while it is held in the node handle are invalidated, and pointers and references obtained to that element before it was extracted become valid.(since C++17)

Reply:
But the point is to use the `unordered_map` so that you get the better look up performance (`Constant on average, worst case linear in the size of the container. ` for `unordered_map` vs `Logarithmic in the size of the container. ` for `map`), and in that case I interpret it to mean that rehashing will invalidate them:

> If after the operation the new number of elements is greater than old [max_load_factor()](https://en.cppreference.com/w/cpp/container/unordered_map/max_load_factor) * [bucket_count()](https://en.cppreference.com/w/cpp/container/unordered_map/bucket_count) a rehashing takes place.
If rehashing occurs (due to the insertion), all iterators are invalidated. Otherwise (no rehashing), iterators are not invalidated. If the insertion is successful, pointers and references to the element obtained while it is held in the node handle are invalidated, and pointers and references obtained to that element before it was extracted become valid.(since C++17)

-- https://en.cppreference.com/w/cpp/container/unordered_map/insert

Which is also the conclusion of this SO answer (which I was using to try and confirm my interpretation): https://stackoverflow.com/a/54004916

----

For what it's worth, I considered using `std::map` and its iterators too.
However, the conclusion I came to is that whether it's a vector of iterators (pointers) or a vector of triplets of pointers, the impact is not much different on memory, even at relatively large scales. Instead, I thought preserving the cpu time optimizations was better for the executor, given the feedback we've consistently gotten, but that's a generalization.